aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -22,12 +22,42 @@ using namespace at::sparse;
 
 namespace {
 
-  inline SparseTensor get_result_tensor_for_unary_op(const SparseTensor& input) {
-    if (c10::isIntegralType(input.scalar_type(), /*includeBool=*/true)) {
+  inline SparseTensor get_result_tensor_for_unary_op(const SparseTensor& input, bool promotes_integer_to_float = false) {
+    if (c10::isIntegralType(input.scalar_type(), /*includeBool=*/true) && promotes_integer_to_float) {
       return at::empty_like(input, input.options().dtype(c10::get_default_dtype()));
     }
     return at::empty_like(input);
   }
+
+  // --------------------------------------------------------------------
+  // helper function for sanity check of input and output tensors
+  // this functions also inits the output tensor
+  // this helper function is for unary ops
+  // --------------------------------------------------------------------
+  static inline void sparse_unary_op_check_and_init(SparseTensor& result, const SparseTensor& self,
+      bool promotes_integer_to_float = false, bool is_linear_map = false) {
+    TORCH_CHECK(result.is_sparse(), "Tensor should be sparse");
+    TORCH_CHECK(self.is_sparse(), "Tensor should be sparse");
+    TORCH_CHECK(self.get_device() == result.get_device(),
+        "Both tensors should be on the same device");
+
+    if (promotes_integer_to_float){
```

Comment:
Let's tweak this check slightly for now. Let me try to write some code for the check here:

```
if (!promotes_integer_to_float) {
  TORCH_CHECK(result.scalar_type() == self.scalar_type(), "the result's dtype must be the same as the input's dtype");
} else {
  if (c10::isIntegralType(self.scalar_type(), /*includeBool=*/true)) {
    TORCH_CHECK(result.scalar_type() == c10::get_default_dtype(), "the result's dtype must be ", c10::get_default_dtype());
  } else {
    TORCH_CHECK(result.scalar_type() == self.scalar_type(), "the result's dtype must be the same as the input's dtype");
  }
}
```

You may want to tweak this (and maybe there's a more elegant way to test the conditions to avoid the redundant TORCH_CHECK), but for now let's just restrict the out= dtype to be input's dtype, or the dtype the input is promoted to.

Reply:
@mruberry @pearu - Have incorporated the above changes